File: templates/heavy-light/edges/HldLazySegTreeEdges.py
# =============================================================================
# 0-INDEXED HLD WITH EDGE-WEIGHTED VARIANT, LAZY PROPAGATION
# =============================================================================
# N = number of nodes, root is node 0
# edges = [(a, b, edgeVal), ...] where edgeVal is the edge's value
# base(rawVal) -> stored data
# combine(data1, data2) -> aggregated data
# applyLazy(lazy, data, segLen) -> data after applying lazy over segLen positions
#   segLen is # of underlying positions covered by data, I just exposed it here
# composeLazies(oldLazy, newLazy) -> combined lazy
#
# Each edge's value is pushed onto its DEEPER endpoint (the child).
# The root has no edge above it, so its slot is unused (None).
# Path/subtree updates and queries always SKIP the relevant LCA / root slot,
# so applyLazy is never invoked on a None data slot.
#
#
# METHODS
# -------
# pointSet(node, newEdgeVal)
#   Sets the value of the edge ABOVE `node`. No-op if node is the root.
#   Takes in the RAW edge data, so for instance a letter if edges have letters
#
# pointApply(node, lazyVal)
#   Applies lazyVal to the edge above `node`. No-op if node is the root.
#
# edgeSet(a, b, newEdgeVal)
#   Sets the value of the edge between adjacent nodes a and b.
#   a and b MUST be adjacent (one is parent of the other).
#
# edgeApply(a, b, lazyVal)
#   Applies lazyVal to the edge between adjacent nodes a and b.
#
# pathApply(a, b, lazyVal)
#   Applies lazyVal to all EDGES on path a..b. No-op if a == b.
#
# subtreeApply(node, lazyVal)
#   Applies lazyVal to all edges from node down to all its descendants.
#   The edge ABOVE node itself is NOT touched.
#
# pathQuery(a, b) -> EdgeData
#   Aggregated data over the EDGES on path a..b.
#   CALLER MUST ENSURE a != b (empty path returns None).
#
# subtreeQuery(node) -> EdgeData
#   Aggregated data over edges from node down to all its descendants.
#   The edge ABOVE node itself is NOT included.
#   CALLER MUST ENSURE node has at least one descendant (leaf returns None).
#
# we don't need a baseLazy() or anything like that, because initially all lazy tags are None,
# and if we apply a lazy to that it just uses the new lazy entirely
#
#
# EXAMPLE: path sum over edge weights with range-add updates
# ----------------------------------------------------------
#   def base(v):
#       return v
#   def combine(a, b):
#       return a + b
#   def applyLazy(lz, d, segLen):
#       return d + lz * segLen   # add lz to each position; sum shifts by lz * segLen
#   def composeLazies(oldL, newL):
#       return oldL + newL
#
#   edges = [
#       (0, 1, 5),
#       (1, 2, 3),
#       (0, 3, 7),
#   ]
#
#   hld = LazyHLDEdge(n, edges, base, combine, applyLazy, composeLazies)
#   hld.pathApply(2, 3, 10)                # add 10 to every edge on path 2..3
#   total = hld.pathQuery(2, 3)
# =============================================================================

class LazyHLDEdge:
    def __init__(self, n, edges, base, combine, applyLazy, composeLazies):
        self.n = n
        self.base = base
        self.combine = combine
        self.applyLazy = applyLazy
        self.composeLazies = composeLazies
        self.adj = [[] for _ in range(n)]
        for a, b, edgeVal in edges:
            self.adj[a].append((b, edgeVal))
            self.adj[b].append((a, edgeVal))
        self.par = [-1] * n
        self.depth = [0] * n
        self.sz = [1] * n
        self.heavy = [-1] * n
        self.head = [0] * n
        self.pos = [0] * n
        self.edgeAbove = [None] * n  # value of edge above each node; None for root
        self._dfsInit()
        self._dfsDecompose()
        self.segN = 1
        while self.segN < max(n, 1):
            self.segN <<= 1
        self.LOG = max(1, self.segN.bit_length() - 1)
        self.tree = [None] * (2 * self.segN)
        self.lazy = [None] * self.segN
        for i in range(n):
            if self.edgeAbove[i] is not None:
                self.tree[self.segN + self.pos[i]] = base(self.edgeAbove[i])
        for i in range(self.segN - 1, 0, -1):
            self.tree[i] = self._combineOpt(self.tree[2*i], self.tree[2*i+1])

    def _combineOpt(self, a, b):
        if a is None: return b
        if b is None: return a
        return self.combine(a, b)

    def _dfsInit(self):
        order = []
        stack = [(0, -1, 0)]
        while stack:
            node, parent, d = stack.pop()
            self.par[node] = parent
            self.depth[node] = d
            order.append(node)
            for nxt, edgeVal in self.adj[node]:
                if nxt != parent:
                    self.edgeAbove[nxt] = edgeVal
                    stack.append((nxt, node, d + 1))
        for node in reversed(order):
            best = 0
            for nxt, edgeVal in self.adj[node]:
                if nxt == self.par[node]: continue
                self.sz[node] += self.sz[nxt]
                if self.sz[nxt] > best:
                    best = self.sz[nxt]
                    self.heavy[node] = nxt

    def _dfsDecompose(self):
        timer = 0
        stack = [(0, 0)]
        while stack:
            node, h = stack.pop()
            self.head[node] = h
            self.pos[node] = timer
            timer += 1
            for nxt, edgeVal in self.adj[node]:
                if nxt == self.par[node] or nxt == self.heavy[node]:
                    continue
                stack.append((nxt, nxt))
            if self.heavy[node] != -1:
                stack.append((self.heavy[node], h))

    def _segLenAt(self, p):
        return self.segN >> (p.bit_length() - 1)

    def _apply(self, p, lazyVal, segLen):
        self.tree[p] = self.applyLazy(lazyVal, self.tree[p], segLen)
        if p < self.segN:
            if self.lazy[p] is None:
                self.lazy[p] = lazyVal
            else:
                self.lazy[p] = self.composeLazies(self.lazy[p], lazyVal)

    def _push(self, p):
        if p < self.segN and self.lazy[p] is not None:
            childSegLen = self._segLenAt(2*p)
            self._apply(2*p, self.lazy[p], childSegLen)
            self._apply(2*p+1, self.lazy[p], childSegLen)
            self.lazy[p] = None

    def _pushTo(self, p):
        for s in range(self.LOG, 0, -1):
            self._push(p >> s)

    def _pullFrom(self, p):
        p >>= 1
        while p:
            newVal = self._combineOpt(self.tree[2*p], self.tree[2*p+1])
            if self.lazy[p] is not None:
                newVal = self.applyLazy(self.lazy[p], newVal, self._segLenAt(p))
            self.tree[p] = newVal
            p >>= 1

    def _segUpdate(self, l, r, lazyVal):
        if l > r: return
        l += self.segN
        r += self.segN + 1
        l0, r0 = l, r
        self._pushTo(l0)
        self._pushTo(r0 - 1)
        segLen = 1
        ll, rr = l, r
        while ll < rr:
            if ll & 1:
                self._apply(ll, lazyVal, segLen)
                ll += 1
            if rr & 1:
                rr -= 1
                self._apply(rr, lazyVal, segLen)
            ll >>= 1
            rr >>= 1
            segLen <<= 1
        self._pullFrom(l0)
        self._pullFrom(r0 - 1)

    def _segQuery(self, l, r):
        if l > r: return None
        l += self.segN
        r += self.segN + 1
        self._pushTo(l)
        self._pushTo(r - 1)
        resL = None
        resR = None
        while l < r:
            if l & 1:
                resL = self._combineOpt(resL, self.tree[l])
                l += 1
            if r & 1:
                r -= 1
                resR = self._combineOpt(self.tree[r], resR)
            l >>= 1
            r >>= 1
        return self._combineOpt(resL, resR)

    def _segPointSet(self, idx, newData):
        p = idx + self.segN
        self._pushTo(p)
        self.tree[p] = newData
        p >>= 1
        while p:
            newVal = self._combineOpt(self.tree[2*p], self.tree[2*p+1])
            if self.lazy[p] is not None:
                newVal = self.applyLazy(self.lazy[p], newVal, self._segLenAt(p))
            self.tree[p] = newVal
            p >>= 1

    def pointSet(self, node, newEdgeVal):
        if node == 0: return  # root has no edge above
        self._segPointSet(self.pos[node], self.base(newEdgeVal))

    def pointApply(self, node, lazyVal):
        if node == 0: return
        p = self.pos[node]
        self._segUpdate(p, p, lazyVal)

    # a and b must be adjacent in the tree; the deeper one is the child whose slot stores this edge
    def edgeSet(self, a, b, newEdgeVal):
        child = a if self.depth[a] > self.depth[b] else b
        self.pointSet(child, newEdgeVal)

    def edgeApply(self, a, b, lazyVal):
        child = a if self.depth[a] > self.depth[b] else b
        self.pointApply(child, lazyVal)

    def pathApply(self, a, b, lazyVal):
        while self.head[a] != self.head[b]:
            if self.depth[self.head[a]] < self.depth[self.head[b]]:
                a, b = b, a
            self._segUpdate(self.pos[self.head[a]], self.pos[a], lazyVal)
            a = self.par[self.head[a]]
        if self.depth[a] > self.depth[b]:
            a, b = b, a
        # a is the LCA; skip its slot by updating [pos[a]+1, pos[b]]
        self._segUpdate(self.pos[a] + 1, self.pos[b], lazyVal)

    def subtreeApply(self, node, lazyVal):
        self._segUpdate(self.pos[node] + 1, self.pos[node] + self.sz[node] - 1, lazyVal)

    # Path query over EDGES from a to b. Skips the LCA's slot since that represents
    # the edge ABOVE the LCA, which is not on the a-b path.
    # Caller must ensure a != b (else returns None).
    def pathQuery(self, a, b):
        res = None
        while self.head[a] != self.head[b]:
            if self.depth[self.head[a]] < self.depth[self.head[b]]:
                a, b = b, a
            chain = self._segQuery(self.pos[self.head[a]], self.pos[a])
            res = self._combineOpt(res, chain)
            a = self.par[self.head[a]]
        if self.depth[a] > self.depth[b]:
            a, b = b, a
        last = self._segQuery(self.pos[a] + 1, self.pos[b])
        return self._combineOpt(res, last)

    # subtree edge query: edges from node down to all its descendants
    # = all stored slots in subtree EXCEPT node's own (which is the edge above node, not in subtree)
    # Caller must ensure node has at least one descendant (else returns None).
    def subtreeQuery(self, node):
        return self._segQuery(self.pos[node] + 1, self.pos[node] + self.sz[node] - 1)

File: templates/heavy-light/edges/lazy/HldLazySegTreeEdges.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;


// EXAMPLE
// using RawEdgeVal = int; // the raw edge value input, could also be a struct

// // Accumulated edge data, for instance total edge weight across edges, could be a `using` too instead of struct
// struct EdgeData {
//     int total;
// };

// // Maps the raw edge value to our EdgeData
// auto base = [](RawEdgeVal v) -> EdgeData {
//     return EdgeData{v};
// };

// // How to aggregate two EdgeData nodes
// auto mergeFn = [](EdgeData A, EdgeData B) -> EdgeData {
//     return EdgeData{A.total + B.total};
// };

// // The lazy data each node stores, could be a `using` too
// struct LazyTag {
//     int pendingAdd;
// };

// // How to apply a lazy update to an EdgeData
// auto applyLazy = [](LazyTag lazy, EdgeData edgeData) -> EdgeData {
//     return EdgeData{edgeData.total + lazy.pendingAdd};
// };

// // How to combine two lazies
// auto composeLazies = [](LazyTag a, LazyTag b) -> LazyTag {
//     return LazyTag{a.pendingAdd + b.pendingAdd};
// };


// TO CREATE THE HLD (O(n) build time)
// RawEdgeVal, EdgeData, LazyTag are the types
// edges should be vector<pair<int,int>> with node labels (either 0...n-1, or 1...n)
// vals[node] is the edge ABOVE that node, if we are using root=1, then vals[0] can be whatever dummy value
// 1 at the end is the root, it can be any value in the tree, separate from if the nodes are 0...n-1 or 1...n
// last param is `zeroIndexed`, so false means the nodes are 1...n
// LazyHLDEdge<RawVal, EdgeData, LazyTag> hld(edges, vals, base, mergeFn, applyLazy, composeLazies, 1, false);


// METHODS

// O(logN)
// pointSet(nodeLabel, RawEdgeVal), overwrites the edge above `nodeLabel` with this new raw value

// O(logN)
// pointApply(nodeLabel, LazyTag), directly apply a lazy update to the edge above `nodeLabel`

// O(log^2 N)
// pathApply(a, b, LazyTag), applies this lazy update to the entire path a...b

// O(log^2 N)
// pathQuery(a, b) -> EdgeData, gives us the aggregated data on the path a...b, crashes if a==b

// O(log N)
// subtreeApply(nodeLabel, LazyTag), applies this update to all edges in the subtree of nodeLabel, no-op if a leaf

// O(log N)
// subtreeQuery(nodeLabel) -> EdgeData, gives us the aggregated data of all edges in the subtree, crashes if a leaf

template <typename RawT, typename StoredT, typename LazyT>
struct LazyHLDEdge {
    int n, root, arrSize;
    bool zeroIndexed;
    function<StoredT(RawT)> base;
    function<StoredT(StoredT, StoredT)> combine;
    function<StoredT(LazyT, StoredT)> applyLazy;
    function<LazyT(LazyT, LazyT)> composeLazies;
    vector<vector<int>> adj;
    vector<int> par, depth, sz, heavy, head, pos;
    int segN, LOG;
    vector<optional<StoredT>> tree;
    vector<optional<LazyT>> lazy;

    LazyHLDEdge(
        const vector<pair<int,int>>& edges,
        const vector<RawT>& vals,
        function<StoredT(RawT)> base,
        function<StoredT(StoredT, StoredT)> combine,
        function<StoredT(LazyT, StoredT)> applyLazy,
        function<LazyT(LazyT, LazyT)> composeLazies,
        int root,
        bool zeroIndexed)
        : n((int)edges.size() + 1), root(root),
          arrSize(zeroIndexed ? (int)edges.size() + 1 : (int)edges.size() + 2), zeroIndexed(zeroIndexed),
          base(base), combine(combine), applyLazy(applyLazy), composeLazies(composeLazies),
          adj(arrSize), par(arrSize, -1), depth(arrSize, 0), sz(arrSize, 1),
          heavy(arrSize, -1), head(arrSize, 0), pos(arrSize, 0)
    {
        // root must be a real node for the chosen mode (caught only with -DDEBUG / asserts on)
        assert(zeroIndexed ? (root >= 0 && root < n) : (root >= 1 && root <= n));
        for (auto& [u, v] : edges) {
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        _dfsInit();
        _dfsDecompose();
        segN = 1;
        while (segN < max(n, 1)) segN <<= 1;
        LOG = max(1, __lg(segN));
        tree.assign(2 * segN, nullopt);
        lazy.assign(segN, nullopt);
        int lo = zeroIndexed ? 0 : 1;
        int hi = zeroIndexed ? n : n + 1;   // exclusive
        for (int i = lo; i < hi; i++) {
            if (i == root) tree[segN + pos[i]] = nullopt;   // no edge above root
            else tree[segN + pos[i]] = base(vals[i]);
        }
        for (int i = segN - 1; i >= 1; i--) {
            tree[i] = _combineOpt(tree[2*i], tree[2*i+1]);
        }
    }

    optional<StoredT> _combineOpt(const optional<StoredT>& a, const optional<StoredT>& b) {
        if (!a.has_value()) return b;
        if (!b.has_value()) return a;
        return combine(*a, *b);
    }

    void _dfsInit() {
        vector<int> order;
        order.reserve(n);
        vector<tuple<int,int,int>> stack;
        stack.push_back({root, -1, 0});
        while (!stack.empty()) {
            auto [node, parent, d] = stack.back();
            stack.pop_back();
            par[node] = parent;
            depth[node] = d;
            order.push_back(node);
            for (int nxt : adj[node]) {
                if (nxt != parent) stack.push_back({nxt, node, d + 1});
            }
        }
        for (int i = (int)order.size() - 1; i >= 0; i--) {
            int node = order[i];
            int best = 0;
            for (int nxt : adj[node]) {
                if (nxt == par[node]) continue;
                sz[node] += sz[nxt];
                if (sz[nxt] > best) {
                    best = sz[nxt];
                    heavy[node] = nxt;
                }
            }
        }
    }

    void _dfsDecompose() {
        int timer = 0;
        vector<pair<int,int>> stack;
        stack.push_back({root, root});
        while (!stack.empty()) {
            auto [node, h] = stack.back();
            stack.pop_back();
            head[node] = h;
            pos[node] = timer++;
            for (int nxt : adj[node]) {
                if (nxt == par[node] || nxt == heavy[node]) continue;
                stack.push_back({nxt, nxt});
            }
            if (heavy[node] != -1) {
                stack.push_back({heavy[node], h});
            }
        }
    }

    void _apply(int p, const LazyT& lazyVal) {
        if (!tree[p].has_value()) return;   // identity slot (root's dummy edge)
        tree[p] = applyLazy(lazyVal, *tree[p]);
        if (p < segN) {
            if (!lazy[p].has_value()) lazy[p] = lazyVal;
            else lazy[p] = composeLazies(*lazy[p], lazyVal);
        }
    }

    void _push(int p) {
        if (p < segN && lazy[p].has_value()) {
            _apply(2*p, *lazy[p]);
            _apply(2*p+1, *lazy[p]);
            lazy[p] = nullopt;
        }
    }

    void _pushTo(int p) {
        for (int s = LOG; s >= 1; s--) _push(p >> s);
    }

    void _pullFrom(int p) {
        for (p >>= 1; p > 0; p >>= 1) {
            auto newVal = _combineOpt(tree[2*p], tree[2*p+1]);
            if (lazy[p].has_value() && newVal.has_value()) {
                newVal = applyLazy(*lazy[p], *newVal);
            }
            tree[p] = newVal;
        }
    }

    void _segUpdate(int l, int r, const LazyT& lazyVal) {
        if (l > r) return;
        l += segN;
        r += segN + 1;
        int l0 = l, r0 = r;
        _pushTo(l0);
        _pushTo(r0 - 1);
        int ll = l, rr = r;
        while (ll < rr) {
            if (ll & 1) { _apply(ll, lazyVal); ll++; }
            if (rr & 1) { rr--; _apply(rr, lazyVal); }
            ll >>= 1; rr >>= 1;
        }
        _pullFrom(l0);
        _pullFrom(r0 - 1);
    }

    optional<StoredT> _segQuery(int l, int r) {
        if (l > r) return nullopt;
        l += segN;
        r += segN + 1;
        _pushTo(l);
        _pushTo(r - 1);
        optional<StoredT> resL = nullopt, resR = nullopt;
        while (l < r) {
            if (l & 1) { resL = _combineOpt(resL, tree[l]); l++; }
            if (r & 1) { r--; resR = _combineOpt(tree[r], resR); }
            l >>= 1; r >>= 1;
        }
        return _combineOpt(resL, resR);
    }

    void _segPointSet(int idx, const optional<StoredT>& newData) {
        int p = idx + segN;
        _pushTo(p);
        tree[p] = newData;
        for (p >>= 1; p > 0; p >>= 1) {
            auto newVal = _combineOpt(tree[2*p], tree[2*p+1]);
            if (lazy[p].has_value() && newVal.has_value()) {
                newVal = applyLazy(*lazy[p], *newVal);
            }
            tree[p] = newVal;
        }
    }

    void pointSet(int node, RawT rawVal) {
        _segPointSet(pos[node], base(rawVal));
    }

    void pointApply(int node, LazyT lazyVal) {
        int p = pos[node];
        _segUpdate(p, p, lazyVal);
    }

    void pathApply(int a, int b, LazyT lazyVal) {
        while (head[a] != head[b]) {
            if (depth[head[a]] < depth[head[b]]) swap(a, b);
            _segUpdate(pos[head[a]], pos[a], lazyVal);
            a = par[head[a]];
        }
        if (depth[a] > depth[b]) swap(a, b);
        if (pos[a] + 1 <= pos[b])
            _segUpdate(pos[a] + 1, pos[b], lazyVal);   // skip LCA's parent edge
    }

    void subtreeApply(int node, LazyT lazyVal) {
        if (sz[node] > 1)
            _segUpdate(pos[node] + 1, pos[node] + sz[node] - 1, lazyVal);
    }

    // precondition: a != b (at least one edge on the path). derefs internally.
    StoredT pathQuery(int a, int b) {
        optional<StoredT> res = nullopt;
        while (head[a] != head[b]) {
            if (depth[head[a]] < depth[head[b]]) swap(a, b);
            res = _combineOpt(res, _segQuery(pos[head[a]], pos[a]));
            a = par[head[a]];
        }
        if (depth[a] > depth[b]) swap(a, b);
        if (pos[a] + 1 <= pos[b])
            res = _combineOpt(res, _segQuery(pos[a] + 1, pos[b]));
        return *res;
    }

    // precondition: node has >= 1 edge in its subtree (not a leaf). derefs internally.
    StoredT subtreeQuery(int node) {
        return *_segQuery(pos[node] + 1, pos[node] + sz[node] - 1);
    }
};
